app/controllers/1/syncs_controller.lua

```diff
@@ -34,94 +55,124 @@ function SyncsController:auth_user()
+    if not redis then
+        self:raise_error(self.error_no_redis)
+    end
+
+    if not is_valid_key_field(self.request.body.username)
+    or not is_valid_field(self.request.body.password) then
+        self:raise_error(self.error_invalid_fields)
+    end
+
+    local user_key = string.format(self.user_key, self.request.body.username)
     local user, err = redis:get(user_key)
     if user == null then
         ok, err = redis:set(user_key, self.request.body.password)
         if not ok then
-            self:raise_error(2000)
+            self:raise_error(self.error_internal)
         else
             return 201, { username = self.request.body.username }
         end
     elseif user then
-        self:raise_error(2002)
+        self:raise_error(self.error_user_exists)
+    else
+        self:raise_error(self.error_internal)
     end
 end
 
 function SyncsController:get_progress()
     local redis = self:getRedis()
+    if not redis then
+        self:raise_error(self.error_no_redis)
+    end
+
     local username = self:authorize()
     if not username then
-        self:raise_error(2001)
-    else
-        local doc = self.params.document
-        if doc then
-            local percent_key = string.format(self.percentage_key, username, doc)
-            local progress_key = string.format(self.progress_key, username, doc)
-            local device_key = string.format(self.device_key, username, doc)
-            local res = {}
-            local percentage, err = redis:get(percent_key)
-            if percentage and percentage ~= null then
-                res.percentage = tonumber(percentage)
-            end
-            local progress, err = redis:get(progress_key)
-            if progress and progress ~= null then
-                res.progress = progress
-            end
-            local device, err = redis:get(device_key)
-            if device and device ~= null then
-                res.device = device
-            end
-            return 200, res
-        else
-            self:raise_error(2003)
-        end
+        self:raise_error(self.error_unauthorized_user)
+    end
+
+    local doc = self.params.document
+    if not is_valid_key_field(doc) then
+        self:raise_error(self.error_document_field_missing)
     end
+
+    local key = string.format(self.doc_key, username, doc)
+    local res = {
+      document = doc,
+    }
+    local results, err = redis:hmget(key,
+                                     self.percentage_field,
+                                     self.progress_field,
+                                     self.device_field,
+                                     self.device_id_field,
+                                     self.timestamp_field)
+    if err then
+        self:raise_error(self.error_internal)
+    end
+
+    if results[1] and results[1] ~= null then
+        res.percentage = tonumber(results[1])
+    end
+    if results[2] and results[2] ~= null then
+        res.progress = results[2]
```

Comment:
The `null` is not a keyword as nil in Lua. It will be treated as a global variable and looked up in the environment table `_G` and evaluated to be `nil`.